scenario/mocking.py

```diff
@@ -688,19 +717,36 @@ def _layers(self) -> Dict[str, pebble.Layer]:
     def _service_status(self) -> Dict[str, pebble.ServiceStatus]:
         return self._container.service_status
 
-    def exec(self, *args, **kwargs):  # noqa: U100 type: ignore
-        cmd = tuple(args[0])
-        out = self._container.exec_mock.get(cmd)
+    # Based on a method of the same name from ops.testing.
+    def _find_exec_handler(self, command) -> Tuple[Optional[str], Optional["Exec"]]:
+        handlers = {
+            tuple(exc.command): (key, exc) for key, exc in self._container.execs.items()
+        }
+        for prefix_len in reversed(range(len(command) + 1)):
+            command_prefix = tuple(command[:prefix_len])
+            if command_prefix in handlers:
+                return handlers[command_prefix]
+        return None, None
+
+    def exec(self, command, **kwargs):  # noqa: U100 type: ignore
+        exec_id, out = self._find_exec_handler(command)
```

Comment:
I get quite itchy from functions that return tuples of untyped values.
I'd say, either A) make it clear from the function name what it's returning (ugly)

```suggestion
        exec_id, exec = self._get_exec_id_and_exec(command)
```

or B)  split it in two calls (better)

```python
exec_id = self._find_exec_id(command)
exec = self._execs[exec_id]
```